ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -910,6 +936,43 @@ func updateStatusToSuspendingIfNeeded(ctx context.Context, rayJob *rayv1.RayJob)
 	return true
 }
 
+func (r *RayJobReconciler) checkSubmitterAndUpdateStatusIfNeeded(ctx context.Context, rayJob *rayv1.RayJob, submissionMode rayv1.JobSubmissionMode) (shouldUpdate, isK8sJobFinished bool, err error) {
+	switch submissionMode {
+	case rayv1.SidecarMode:
+		shouldUpdate = r.checkSidecarContainerAndUpdateStatusIfNeeded(ctx, rayJob)
+		// Sidecar mode doesn't use k8s job, so isK8sJobFinished is always false
+		isK8sJobFinished = false
+		err = nil
+		return
+	case rayv1.K8sJobMode:
+		logger := ctrl.LoggerFrom(ctx)
+		job := &batchv1.Job{}
+		// If the submitting Kubernetes Job reaches the backoff limit, transition the status to `Complete` or `Failed`.
+		// This is because, beyond this point, it becomes impossible for the submitter to submit any further Ray jobs.
+		// For light-weight mode, we don't transition the status to `Complete` or `Failed` based on the number of failed
+		// requests. Instead, users can use the `ActiveDeadlineSeconds` to ensure that the RayJob in the light-weight
+		// mode is not stuck in the `Running` status indefinitely.
+		namespacedName := common.RayJobK8sJobNamespacedName(rayJob)
+		if err = r.Client.Get(ctx, namespacedName, job); err != nil {
+			logger.Error(err, "Failed to get the submitter Kubernetes Job for RayJob", "NamespacedName", namespacedName)
+			shouldUpdate = false
+			isK8sJobFinished = false
+			return
+		}
+		shouldUpdate = checkK8sJobAndUpdateStatusIfNeeded(ctx, rayJob, job)
+		_, isK8sJobFinished = utils.IsJobFinished(job)
+		err = nil
+		return
+	default:
+		logger := ctrl.LoggerFrom(ctx)
```

Comment:
This means the KubeRay logic is wrong. I’m considering using panic instead. It’s not very common to use panic in an operator, but in my opinion, it’s still better than letting it keep reconciling in a wrong state.

In addition, this is a system error instead of application error, but the error message "you can only call ..." looks like application error.